src/test/librbd/test_librbd.cc

```diff
@@ -3099,6 +3099,192 @@ TEST_F(TestLibRBD, TestAioCompareAndWriteStripeUnitSuccess)
+    {.iov_base = &small_buffer[0], .iov_len = 4},
+    {.iov_base = &small_buffer[4], .iov_len = 4}
+  };
+
+  // should fail because compare iovec len cannot be different to write iovec len
+  rbd_completion_t comp;
+  rbd_aio_create_completion(NULL, NULL, &comp);
+  uint64_t mismatch_off = 0;
+  int ret = rbd_aio_compare_and_writev(image, off,
+                                       small_iovs, /* cmp_iovs */
+                                       sizeof(small_iovs) / sizeof(struct iovec),
+                                       write_iovs,
+                                       sizeof(write_iovs) / sizeof(struct iovec),
+                                       comp,
+                                       &mismatch_off, 0);
+  ASSERT_EQ(-EINVAL, ret);
+  ASSERT_EQ(0U, mismatch_off);
+  rbd_aio_release(comp);
+
+  // check nothing was written
+  std::string read_buffer(cmp_buffer.length(), '1');
+  ssize_t read = rbd_read(image, off, read_buffer.length(), read_buffer.data());
+  ASSERT_EQ(read_buffer.length(), read);
+  ASSERT_EQ(cmp_buffer, read_buffer);
+
+  ASSERT_PASSED(validate_object_map, image);
+  ASSERT_EQ(0, rbd_close(image));
+
+  rados_ioctx_destroy(ioctx);
+}
+
+TEST_F(TestLibRBD, TestAioCompareAndWritevMismatch)
+{
+  rados_ioctx_t ioctx;
+  rados_ioctx_create(_cluster, m_pool_name.c_str(), &ioctx);
+
+  rbd_image_t image;
+  int order = 0;
+  std::string name = get_temp_image_name();
+  uint64_t size = 20 << 20; /* 20MiB */
+  off_t off = 512;
+
+  ASSERT_EQ(0, create_image(ioctx, name.c_str(), size, &order));
+  ASSERT_EQ(0, rbd_open(ioctx, name.c_str(), &image, NULL));
+
+  std::string cmp_buffer("This is a test");
+  int cmp_len = cmp_buffer.length();
+
+  std::string write_buffer("Write this !!!");
+  struct iovec write_iovs[] = {
+    {.iov_base = &write_buffer[0], .iov_len = 6},
+    {.iov_base = &write_buffer[6], .iov_len = 5},
+    {.iov_base = &write_buffer[11], .iov_len = 3}
+  };
+  int write_iovcnt = sizeof(write_iovs) / sizeof(struct iovec);
+
+  std::string mismatch_buffer("This will fail");
+  struct iovec mismatch_iovs[] = {
+    {.iov_base = &mismatch_buffer[0], .iov_len = 5},
+    {.iov_base = &mismatch_buffer[5], .iov_len = 5},
+    {.iov_base = &mismatch_buffer[10], .iov_len = 4}
+  };
+  int mismatch_iovcnt = sizeof(mismatch_iovs) / sizeof(struct iovec);
+
+  ASSERT_EQ(cmp_len, rbd_write(image, off, cmp_len, cmp_buffer.data()));
+
+  // this should execute the compare but fail because of mismatch
+  rbd_completion_t comp;
+  rbd_aio_create_completion(NULL, NULL, &comp);
+  uint64_t mismatch_off = 0;
+  int ret = rbd_aio_compare_and_writev(image, off,
+                                       mismatch_iovs, /* cmp_iovs */
+                                       mismatch_iovcnt,
+                                       write_iovs,
+                                       write_iovcnt,
+                                       comp,
+                                       &mismatch_off, 0);
```

Comment:
Get rid of `mismatch_iovcnt` and `write_iovcnt`:
```suggestion
                                       std::size(mismatch_iovs),
                                       write_iovs, std::size(write_iovs),
                                       comp, &mismatch_off, 0);
```